crates/task-impls/src/quorum_proposal.rs

```diff
@@ -80,32 +66,184 @@ struct ProposalDependencyHandle<TYPES: NodeType> {
     sender: Sender<Arc<HotShotEvent<TYPES>>>,
 
     /// Reference to consensus. The replica will require a write lock on this.
-    #[allow(dead_code)]
     consensus: Arc<RwLock<Consensus<TYPES>>>,
+
+    /// Membership for Quorum Certs/votes
+    quorum_membership: Arc<TYPES::Membership>,
+
+    /// Our public key
+    public_key: TYPES::SignatureKey,
+
+    /// Our Private Key
+    private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+
+    /// last Upgrade Certificate this node formed
+    upgrade_cert: Option<UpgradeCertificate<TYPES>>,
+}
+
+impl<TYPES: NodeType> ProposalDependencyHandle<TYPES> {
+    /// Checks if we have somehow initiated a proposal while not being the leader of the next
+    /// `view`.
+    fn check_leader(&self, view: TYPES::Time) -> Result<()> {
+        if self.quorum_membership.get_leader(view) != self.public_key {
+            // This is expected for view 1, so skipping the logging.
+            if view != TYPES::Time::new(1) {
+                bail!(
+                    "Somehow we formed a QC but are not the leader for the next view {:?}",
+                    view
+                );
+            }
+        }
```

Comment:
I think this could be `ensure!(self.quorum_membership.get_leader(view) != self.public_key, "Somehow...", view);`.

You could `&& view != TYPES::Time::new(1)`, but I'm ambivalent. At least personally, I don't care if view 1 has unnecessary logs, it's just a single view and should be fixed relatively soon: #2873).